### PR TITLE
ARROW-13200: [R] Add binding for case_when()

### PR DESCRIPTION
**Pull request by ianmcook**
Adds support for `case_when()` in dplyr verbs. I followed the example of `dbplyr::case_when()`, which is much simpler and more self-contained than `dplyr::case_when()`.

### CODE REVIEW

**Comment by ianmcook** [comment]
Let's merge #10724 before we merge this. I will rebase and resolve the conflicts after #10724 is merged.

**Comment by ianmcook** [comment]
@github-actions crossbow submit test-r-without-arrow test-r-minimal-build

File: r/R/dplyr-functions.R [hunk 1]
```diff
@@ -169,7 +169,7 @@ nse_funcs$is.integer64 <- function(x) {
   x$type_id() == Type["INT64"]
 }
 nse_funcs$is.logical <- function(x) {
-  x$type_id() == Type["BOOL"]
+  is.logical(x) || (inherits(x, "Expression") && x$type_id() == Type["BOOL"])
```
Comment (ianmcook) [inline r/R/dplyr-functions.R:172]: This is a dup of a change made in #10724

File: r/R/dplyr-functions.R [hunk 2]
```diff
@@ -651,3 +651,47 @@ nse_funcs$log <- function(x, base = exp(1)) {
...
+      abort("Left side of each formula in case_when() must be a logical expression")
+    }
+    # TODO: remove these checks after the case_when kernel supports variable-width
+    # types (ARROW-13222)
+    has_bad_r_type <- inherits(value[[i]], c("character", "raw", "list", "factor"))
+    has_bad_arrow_type <- inherits(value[[i]], "Expression") &&
+      value[[i]]$type_id() %in% Type[c(
+      "STRING", "BINARY", "LIST", "MAP", "STRUCT", "SPARSE_UNION",
+      "DENSE_UNION", "DICTIONARY", "EXTENSION", "FIXED_SIZE_LIST",
+      "LARGE_STRING", "LARGE_BINARY", "LARGE_LIST")]
```
Comment (ianmcook) [inline r/R/dplyr-functions.R:681]: Some of these types are not testable in R but are obviously variable-width. For the ones that are testable in R, I tested them to confirm they don't yet work with the `case_when` kernel, and I tested all the others to confirm they do work.

File: r/R/dplyr-functions.R [hunk 3]
```diff
@@ -651,3 +651,47 @@ nse_funcs$log <- function(x, base = exp(1)) {
...
+    if (!inherits(f, "formula")) {
+      abort("Each argument to case_when() must be a two-sided formula")
+    }
+    query[[i]] <- arrow_eval(f[[2]], mask)
+    value[[i]] <- arrow_eval(f[[3]], mask)
+    if (!nse_funcs$is.logical(query[[i]])) {
+      abort("Left side of each formula in case_when() must be a logical expression")
+    }
+    # TODO: remove these checks after the case_when kernel supports variable-width
+    # types (ARROW-13222)
```
Comment (nealrichardson) [inline r/R/dplyr-functions.R:675]: Can we just let the Arrow C++ library error out if it doesn't support the types? Is our error message that much better? Seems like extra work to try to keep this validation in sync with the C++ library.

File: r/tests/testthat/test-dplyr.R [hunk 4]
```diff
@@ -1063,4 +1063,123 @@ test_that("trig functions", {
...
+      )) %>%
+      collect(),
+    tbl %>%
+      transmute(
+        cw = ifelse(is.na(fct), int, ifelse(is.na(chr), dbl, 1e10))
+      )
+  )
+
+  # expected errors (which are caught by abandon_ship() and changed to warnings)
+  # TODO: Find a way to test these directly without abandon_ship() interfering
```
Comment (nealrichardson) [inline r/tests/testthat/test-dplyr.R:1117]: `expect_error(nse_funcs$case_when(), "case_when")`?

Comment (ianmcook) [reply r/tests/testthat/test-dplyr.R:1278]: For the zero-cases error, yes, that would work. But for the other errors I don't think it would.

**Comment by nealrichardson** [comment]
I've merged the news PR, so after you rebase please add a bullet in the new dplyr functions section for this (noting any type limitations) and also note if_else there too (which I think doesn't have those same type limitations?)

File: r/R/dplyr-functions.R (hunk 3 above)
Comment (ianmcook) [reply r/R/dplyr-functions.R]: Yep, agreed, there's really not much (any?) benefit to having these checks here. Removed in 2ee3c8c5909857452d89a722459127ddde723c05.

File: r/tests/testthat/test-dplyr.R (hunk 4 above)
Comment (ianmcook) [reply r/tests/testthat/test-dplyr.R:1278]: I tried several approaches for testing these errors directly, but I ended up with various errors and rlang data mask warnings. IMO: let's leave this as is with the TODO to revisit later.

Comment (nealrichardson) [reply r/tests/testthat/test-dplyr.R:1278]:
```suggestion
  # TODO: ARROW-13370 to test the messages better
```

File: r/src/compute.cpp [hunk 5]
```diff
@@ -241,6 +241,13 @@ std::shared_ptr<arrow::compute::FunctionOptions> make_compute_options(
     return out;
   }
 
+  if (func_name == "make_struct") {
+    using Options = arrow::compute::MakeStructOptions;
+    // TODO: accept `field_nullability` and `field_metadata` options
```
Comment (nealrichardson) [inline r/src/compute.cpp:246]:
```suggestion
    // TODO (ARROW-13371): accept `field_nullability` and `field_metadata` options
```

**Review by nealrichardson — APPROVED** [review]
Looks great!

**Pull request closed without merging**